### PR TITLE
(feat) Synchronously load most extensions and pages

### PR DESCRIPTION
**Pull request by ibacher**
## Requirements
- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.

#### For changes to apps
- [ ]  My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide) and **design documentation**.

#### If applicable
- [ ] My work includes tests or is validated by existing tests.
- [ ] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary

This ensures that all components are in the main chunk of the apps in esm-core. This reduces the number of JS requests by about 13, though I think there are more significant gains to be had by adopting this in the patient-chart and patient-management repos, especially.

At least for me, the primary navigation is, expected, a little snappier under this change.

## Other

Added bonus: this adds types for the Webpack config.

### CODE REVIEW

File: packages/tooling/webpack-config/src/index.ts [hunk 1]
```diff
@@ -245,6 +266,17 @@ export default (
     performance: {
       hints: mode === production && "warning",
     },
+    optimization: merge(
+      {
+        // The defaults for both of these are 30; however, due to the modular nature of
+        // the frontend, we want each app to produce substantially
```
Comment (denniskigen) [inline packages/tooling/webpack-config/src/index.ts:272]: Looks like this comment isn't complete.

File: packages/tooling/webpack-config/src/index.ts [hunk 2]
```diff
@@ -148,7 +166,10 @@ export default (
     main,
     types,
   } = require(resolve(root, "package.json"));
-  const mode = argv.mode || process.env.NODE_ENV || "development";
+  // this typing is provably incorrect, but actually works
```
Comment (denniskigen) [inline packages/tooling/webpack-config/src/index.ts:169]:
```suggestion
  // this typing is probably incorrect, but actually works
```

File: packages/tooling/webpack-config/src/index.ts (hunk 1 above)
Comment (ibacher) [reply packages/tooling/webpack-config/src/index.ts:272]: Right I got involved with writing an override for it and forgot to go back.

File: packages/tooling/webpack-config/src/index.ts (hunk 2 above)
Comment (ibacher) [reply packages/tooling/webpack-config/src/index.ts:169]: I actually meant "provably" here. The value isn't guaranteed to be one of "none", "production", or "development", but I'd rather leave that validation to Webpack than figure it out here, so I'm lying to Typescript 😁.

File: packages/tooling/webpack-config/src/index.ts [hunk 3]
```diff
@@ -40,14 +40,24 @@ import ForkTsCheckerWebpackPlugin from "fork-ts-checker-webpack-plugin";
...
+  DefinePlugin,
+  container,
+  type ModuleOptions,
+  type WebpackOptionsNormalized as WebpackConfiguration,
+  type RuleSetRule,
+} from "webpack";
 import { BundleAnalyzerPlugin } from "webpack-bundle-analyzer";
 import { StatsWriterPlugin } from "webpack-stats-plugin";
 // eslint-disable-next-line no-restricted-imports
 import { merge, mergeWith, isArray } from "lodash";
```
Comment (denniskigen) [inline packages/tooling/webpack-config/src/index.ts:53]: Worth fixing this?

```suggestion
import merge from "lodash-es/merge";
import mergeWith from "lodash-es/mergeWith";
import isArray from "lodash-es/isArray";
```

**Review by denniskigen — APPROVED** [review]
Great work, @ibacher! Thanks!

File: packages/tooling/webpack-config/src/index.ts (hunk 3 above)
Comment (ibacher) [reply packages/tooling/webpack-config/src/index.ts:53]: In the webpack-config? Nah... We don't actually include it in a bundle, so the fact that `lodash` doesn't really tree-shake here is a non-issue. We're basically just using it as a NodeJS library.

**Pull request merged**